src/librawspeed/decoders/IiqDecoder.cpp

```diff
@@ -242,6 +260,153 @@ void IiqDecoder::DecodePhaseOneC(const std::vector<IiqStrip>& strips,
+    const uint32 len = entries.getU32();
+    const uint32 offset = entries.getU32();
+
+    switch (tag) {
+    case 0x431:
+      if (!quadrant_multiplier_applied) {
+        CorrectQuadrantMultipliersCombined(meta_data.getSubStream(offset, len),
+                                           split_row, split_col);
+      }
+      quadrant_multiplier_applied = true;
+      break;
+    default:
+      break;
+    }
+  }
+}
+
+// This function defines a correction that compensates for the fact
+// that IIQ files may come from a camera with multiple (four, in this
+// case) sensors combined into a single "sensor."  Because the
+// different sensors may have slightly different responses, we need to
+// multiply the pixels in each by a correction factor to ensure that
+// they blend together smoothly.  The correction factor is not a
+// single multiplier, but a curve defined by seven control points.
+// Each curve's control points share the same seven X-coordinates.
+void IiqDecoder::CorrectQuadrantMultipliersCombined(ByteStream data,
+                                                    uint32 split_row,
+                                                    uint32 split_col) {
+
+  // The curves should all include (0, 0) and (65535, 65535), so the
+  // first and last points are predefined and the middle seven are
+  // read from the file
+  uint32 shared_x_coords[9] = {0, 0, 0, 0, 0, 0, 0, 0, 65535};
+  for (int i = 0; i < 7; i++)
+    shared_x_coords[i + 1] = data.getU32();
+
+  uint32 curve_control_points_y[2][2][9];
+  for (int quadRow = 0; quadRow < 2; quadRow++) {
+    for (int quadCol = 0; quadCol < 2; quadCol++) {
+      curve_control_points_y[quadCol][quadRow][0] = 0;
+      curve_control_points_y[quadCol][quadRow][8] = 65535;
+      for (int i = 0; i < 7; i++)
+        // These multipliers are expressed in ten-thousandths in the
+        // file
+        curve_control_points_y[quadRow][quadCol][i + 1] =
+            (data.getU32() * shared_x_coords[i + 1]) / 10000;
+    }
+  }
+
+  ushort16 curve[65536];
+  for (int quadRow = 0; quadRow < 2; quadRow++) {
+    for (int quadCol = 0; quadCol < 2; quadCol++) {
+      calculateCurve(shared_x_coords, curve_control_points_y[quadRow][quadCol],
+                     9, curve);
+
+      int row_start = quadRow == 0 ? 0 : split_row;
+      int row_end = quadRow == 0 ? split_row : mRaw->dim.y;
+      int col_start = quadCol == 0 ? 0 : split_col;
+      int col_end = quadCol == 0 ? split_col : mRaw->dim.x;
+
+      for (int row = row_start; row < row_end; row++) {
+        for (int col = col_start; col < col_end; col++) {
+          ushort16* pixel =
+              reinterpret_cast<ushort16*>(mRaw->getData(col, row));
+          *pixel = curve[*pixel];
+        }
+      }
+    }
+  }
+}
+
+// This is an implementation of the cubic spline algorithm found on
+// Wikipedia at https://en.wikipedia.org/wiki/Spline_(mathematics)
+void IiqDecoder::calculateCurve(const uint32* x_coords, const uint32* y_coords,
```

Comment:
Let's put it into `src/librawspeed/common/Spline.h` i guess.

Reply:
Also *DONT* take pointers like this.

Reply:
Okay, so I'm guessing I should just pass the coords vectors around as values since they're small-is?